hastory-cli/src/Hastory/Cli/OptParse/Types.hs

```diff
@@ -14,59 +20,160 @@ data Instructions =
   deriving (Show)
 
 data Command
-  = CommandGather
-  | CommandGenGatherWrapperScript
-  | CommandListRecentDirs ListRecentDirArgs
-  | CommandChangeDir Int
-  | CommandGenChangeWrapperScript
-  | CommandSuggestAlias
+  = CommandGather GatherFlags
+  | CommandGenGatherWrapperScript GenGatherWrapperScriptFlags
+  | CommandListRecentDirs ListRecentDirFlags
+  | CommandChangeDir ChangeDirFlags
+  | CommandGenChangeWrapperScript GenChangeWrapperScriptFlags
+  | CommandSuggestAlias SuggestAliasFlags
+  deriving (Show, Eq)
+
+data GatherFlags =
+  GatherFlags
+  deriving (Show, Eq)
+
+data GenGatherWrapperScriptFlags =
+  GenGatherWrapperScriptFlags
+  deriving (Show, Eq)
+
+newtype ChangeDirFlags =
+  ChangeDirFlags
+    { changeDirFlagsIdx :: Int
+    }
+  deriving (Show, Eq)
+
+data GenChangeWrapperScriptFlags =
+  GenChangeWrapperScriptFlags
   deriving (Show, Eq)
 
-newtype ListRecentDirArgs =
-  ListRecentDirArgs
+data SuggestAliasFlags =
+  SuggestAliasFlags
+  deriving (Show, Eq)
+
+newtype ListRecentDirFlags =
+  ListRecentDirFlags
     { lrdArgBypassCache :: Maybe Bool
     }
   deriving (Show, Eq)
 
 data Flags =
   Flags
     { flagCacheDir :: Maybe FilePath
-    , flagStorageServer :: Maybe Text
+    , flagConfigFile :: Maybe FilePath
+    , flagStorageServer :: Maybe BaseUrl
     , flagStorageUsername :: Maybe Username
     , flagStoragePassword :: Maybe Text
     }
-  deriving (Show)
+  deriving (Show, Eq)
+
+data Environment =
+  Environment
+    { envCacheDir :: Maybe FilePath
+    , envConfigFile :: Maybe FilePath
+    , envStorageServer :: Maybe BaseUrl
+    , envStorageUsername :: Maybe Username
+    , envStoragePassword :: Maybe Text
+    }
+  deriving (Show, Eq)
 
 data Configuration =
   Configuration
+    { configCacheDir :: Maybe FilePath
+    , configStorageServer :: Maybe BaseUrl
+    , configStorageUsername :: Maybe Username
+    , configStoragePassword :: Maybe Text
+    }
   deriving (Show, Eq)
 
+defaultConfiguration :: Configuration
+defaultConfiguration =
+  Configuration
+    { configCacheDir = Nothing
+    , configStorageServer = Nothing
+    , configStorageUsername = Nothing
+    , configStoragePassword = Nothing
+    }
+
+instance YamlSchema Configuration where
+  yamlSchema = parseEmptyConfigurationFile <|> parseObject
+    where
+      parseEmptyConfigurationFile =
```

Comment:
I don't think you need to be able to deal with an empty configuratino file, I'd suggest removing this.

Reply:
I think I understand. [readConfigFile](https://hackage.haskell.org/package/yamlparse-applicative-0.1.0.1/docs/YamlParse-Applicative.html#v:readConfigFile) will not try to parse an empty file (or missing) file, so there is no reason for us to handle that case.